contracts/javascore/aggregator/src/main/java/relay/aggregator/RelayAggregator.java

```diff
@@ -0,0 +1,305 @@
+        Context.require(existingSign == null, "Signature already exists");
+
+        setSignature(pktID, Context.getCaller(), signature);
+
+        if (signatureThresholdReached(pktID)) {
+            byte[][] sigs = getSignatures(srcNetwork, contractAddress, srcSn);
+            byte[] encodedSigs = serializeSignatures(sigs);
+            PacketAcknowledged(
+                    pkt.getSrcNetwork(),
+                    pkt.getContractAddress(),
+                    pkt.getSrcSn(),
+                    pkt.getSrcHeight(),
+                    pkt.getDstNetwork(),
+                    pkt.getData(),
+                    encodedSigs);
+            removePacket(pktID);
+        }
+    }
+
+    private byte[][] getSignatures(String srcNetwork, String contractAddress, BigInteger srcSn) {
+        String pktID = Packet.createId(srcNetwork, contractAddress, srcSn);
+        DictDB<Address, byte[]> signDict = signatures.at(pktID);
+        ArrayList<byte[]> signatureList = new ArrayList<byte[]>();
+
+        for (int i = 0; i < relayers.size(); i++) {
+            Address relayer = relayers.get(i);
+            byte[] sign = signDict.get(relayer);
+            if (sign != null) {
+                signatureList.add(sign);
+            }
+        }
+
+        byte[][] sigs = new byte[signatureList.size()][];
+        for (int i = 0; i < signatureList.size(); i++) {
+            sigs[i] = signatureList.get(i);
+        }
+        return sigs;
+    }
+
+    protected void setSignature(String pktID, Address addr, byte[] sign) {
+        signatures.at(pktID).set(addr, sign);
+    }
+
+    protected static byte[] serializeSignatures(byte[][] sigs) {
+        ByteArrayObjectWriter w = Context.newByteArrayObjectWriter("RLPn");
+        w.beginList(sigs.length);
+
+        for (byte[] sig : sigs) {
+            w.write(sig);
+        }
+
+        w.end();
+        return w.toByteArray();
+    }
+
+    protected static byte[][] deserializeSignatures(byte[] encodedSigs) {
+        ObjectReader r = Context.newByteArrayObjectReader("RLPn", encodedSigs);
+
+        ArrayList<byte[]> sigList = new ArrayList<>();
+
+        r.beginList();
+        while (r.hasNext()) {
+            sigList.add(r.readByteArray());
+        }
+        r.end();
+
+        byte[][] sigs = new byte[sigList.size()][];
+        for (int i = 0; i < sigList.size(); i++) {
+            sigs[i] = sigList.get(i);
+        }
+
+        return sigs;
+    }
+
+    private void adminOnly() {
+        Context.require(Context.getCaller().equals(admin.get()), "Unauthorized: caller is not the leader relayer");
+    }
+
+    private void relayersOnly() {
+        Address caller = Context.getCaller();
+        Boolean isRelayer = false;
+        for (int i = 0; i < relayers.size(); i++) {
+            Address relayer = relayers.get(i);
+            if (relayer.equals(caller)) {
+                isRelayer = true;
+                break;
+            }
```

Comment:
this could be optimized to  constant time using dict db

Reply:
> this could be optimized to constant time using dict db

Added.